core/trino-main/src/main/java/io/trino/cost/StatsCalculator.java

```diff
@@ -34,10 +34,11 @@ PlanNodeStatsEstimate calculateStats(
             StatsProvider sourceStats,
             Lookup lookup,
             Session session,
-            TypeProvider types);
+            TypeProvider types,
+            TableStatsProvider tableStatsProvider);
```

Comment:
Why do we need a new entity here, the `tableStatsProvider`?
Could `StatsProvider sourceStats` do the job?

Reply:
`StatsProvider sourceStats` is called to get children plan nodes' stats.

`TableStatsProvider tableStatsProvider` is used when calculating the "ultimate" source stats: table scan stats. In fact, all this PR does is to wire in a `CachingTableStatsProvider` for `TableScanStatsRule` to use.

Reply:
You can easily imagine `CachingStatsProvider` that caches stats on per-PlanNode basis. This would be a generally useful implementation, potentially allowing us to cut down some stats calculation cost.

The currently existing `CachingStatsProvider` caches stats in Memo's group. This is useful as well -- the group contains alternative plans (currently always exactly one at a time) that produce same relation. Same relation implies that once the stats are calculated, they are applicable to all group members.

Now, these two concepts are different, but not exclusionary:

- we could have "L2" cache that's on per-PlanNode basis
- then "L1" cache that's based on Memo group

Now the question is, whether per-PlanNode basis and per-TableHandle basis are significantly different.
I think there are not. I don't think we ever have a case where we have two `TableScanNode` that have same `TableHandle`.

Reply:
> Now the question is, whether per-PlanNode basis and per-TableHandle basis are significantly different.

Actually they would be different.

Plan nodes have no equality, they compare by identity. Upon each exit from IterativeOptimizer (exit from Memo), a fully new plan structure is created. This can be improved a bit (eg produce new plan only if anything changed), but still would break identity-based caching, except for root nodes (table scans). Thus, the solution would look more generic, but would not actually be.